### PR TITLE
feat: add decouple and simple test and two_pc

### PR DESCRIPTION
**Pull request by IceCoooola**

### CODE REVIEW

**Comment by IceCoooola** [comment]
upload two_pc

**Comment by shadaj** [comment]
@IceCoooola you will need to rebase because we made a couple of breaking changes. In particular, you can now just call `.self_id()` on a `Cluster` to get that value, which should simplify some of your logic.

**Comment by IceCoooola** [comment]
@shadaj rebased!

File: hydroflow_plus/src/stream.rs [hunk 1]
```diff
@@ -690,6 +690,42 @@ pub(super) fn deserialize_bincode<T: DeserializeOwned>(tagged: bool) -> Pipeline
...
 impl<'a, T, W, N: Location<'a>> Stream<T, W, NoTick, N> {
+    pub fn decouple_process<P2>(
+        self,
+        other: &Process<P2>,
+    ) -> Stream<'a, T, Unbounded, NoTick, Process<P2>>
+    where
+        N: CanSend<Process<P2>, In<T> = T, Out<T> = T>,
+        T: Clone + Serialize + DeserializeOwned,
+    {
+        match self.location_kind {
```
Comment (shadaj) [inline hydroflow_plus/src/stream.rs:701]: Shouldn't need to do this check, because since `Out<T> = T` we know the source must be a process.

File: hydroflow_plus/src/stream.rs [hunk 2]
```diff
@@ -847,8 +883,10 @@ impl<'a, T, W, N: Location<'a>> Stream<T, W, NoTick, N> {
         N: CanSend<'a, Cluster<'a, C2>, In<T> = (u32, T)>,
         T: Clone + Serialize + DeserializeOwned,
     {
-        let ids = other.members();
```
Comment (shadaj) [inline hydroflow_plus/src/stream.rs:850]: Should be able to undo this change.

File: hydroflow_plus_test/src/cluster/simple_cluster.rs [hunk 3]
```diff
@@ -30,6 +52,51 @@ mod tests {
...
+                &cluster2,
+                (0..3)
+                    .map(|_| TrybuildHost::new(deployment.Localhost()))
+                    .collect::<Vec<_>>(),
+            )
+            .deploy(&mut deployment);
+
+        deployment.deploy().await.unwrap();
+
+        deployment.start().await.unwrap();
```
Comment (shadaj) [inline hydroflow_plus_test/src/cluster/simple_cluster.rs:80]: This unit test doesn't actually check anything; we should be inspecting the `stdout` of nodes.

File: hydroflow_plus_test/src/cluster/simple_cluster.rs [hunk 4]
```diff
@@ -30,6 +52,51 @@ mod tests {
...
+        let built = builder.with_default_optimize();
+
+        let _nodes = built
+            .with_process(&process1, TrybuildHost::new(deployment.Localhost()))
+            .with_process(&process2, TrybuildHost::new(deployment.Localhost()))
+            .deploy(&mut deployment);
+
+        deployment.deploy().await.unwrap();
+
+        deployment.start().await.unwrap();
```
Comment (shadaj) [inline hydroflow_plus_test/src/cluster/simple_cluster.rs:98]: Same as above

File: hydroflow_plus_test/src/cluster/two_pc.rs [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+                    .collect::<Vec<_>>(),
+            )
+            .with_process(&client, TrybuildHost::new(deployment.Localhost()))
+            .deploy(&mut deployment);
+        // println!("{:?}", built.ir());
+        deployment.deploy().await.unwrap();
+
+        deployment.start().await.unwrap();
+
+        tokio::signal::ctrl_c().await.unwrap();
```
Comment (shadaj) [inline hydroflow_plus_test/src/cluster/two_pc.rs:119]: This unit test will hang because `Ctrl-C` won't be invoked; we should move this to an example for now.

**Review by shadaj — CHANGES_REQUESTED** [review]
First pass!

File: .gitignore [hunk 6]
```diff
@@ -1,3 +1,4 @@
+.vscode/
```
Comment (shadaj) [inline .gitignore:1]: We shouldn't add this, because we have meaningful configuration checked into the repo.

**Comment by IceCoooola** [comment]
@shadaj fixed! could you review it again? thank you!

File: hydroflow_plus/src/stream.rs [hunk 7]
```diff
@@ -690,6 +690,38 @@ pub(super) fn deserialize_bincode<T: DeserializeOwned>(tagged: bool) -> Pipeline
...
+    {
+        self.send_bincode::<Process<'a, P2>, T>(other)
+    }
+
+    pub fn decouple_cluster<C2>(
+        self,
+        other: &Cluster<'a, C2>,
+    ) -> Stream<N::Out<T>, Unbounded, NoTick, Cluster<'a, C2>>
+    where
+        N: CanSend<'a, Cluster<'a, C2>, In<T> = (u32, T)>,
```
Comment (shadaj) [inline hydroflow_plus/src/stream.rs:709]: We should restrict `Out<T>` here to be `(u32, T)` as well, and then use `send_bincode_interleaved` so that we guarantee that we have the same type on the input and output (the return type should be `Stream<T, ...>`).

**Review by shadaj — CHANGES_REQUESTED** [review]

**Review by shadaj — APPROVED** [review]
Awesome! 🎉
